<h2>Question:</h2>
<h4>Write a program that with a specific word length, and the program will search through a file consisting English words, for sets of words that are that length and are anagrams of each other.</h4>

<h4>Note:</h4> 
<ul>
    <li>The English words are downloaded from <a href="https://github.com/dwyl/english-words">
        https://github.com/dwyl/english-words</a></li>
    <li>Due to space constrain, this page doesn't show the complete output.</li>    
    <li>Below function may be slow because the way List.remove() is implemented. 
        We may implement the function differently in order to make it run faster.</li>   
</ul>    
<br><br><br><br>
<h4>The original question:</h4>
<img src="Question.jpg" width="80%">

In [1]:
# Define the function
def is_anagram(subject, anagram):  
    if len(subject) != len(anagram):
        return False  # Anagram must consist same number of letters.
    
    list2 = list(subject)  # To convert the string into a list.
    for x in anagram:
        try:
            list2.remove(x)  # Remove this letter because it has been "used up".
        except ValueError:
            return False  # It is not anagram if the letter is not found in the list.
    
    # return len(list2) == 0  # <-- This is redundant.
    return True    

In [2]:
def main():
    with open('words_incomplete.txt') as f:
        words = f.read().splitlines()  # f.readlines()

    length2 = 12  # <-- Get this from user input.
    N = len(words)

    for i in range(N):
        if len(words[i]) == length2:
            for j in range(i+1, N):
                if is_anagram(words[i], words[j]):
                    print("{}, {}".format(words[i], words[j]))

In [3]:
print(__name__)
if __name__ == "__main__":
    import cProfile
    cProfile.run("main()")

__main__
acclimatiser, cameralistic
accouterment, accoutrement
acherontical, anchoretical
acidophilous, aphidicolous
alterability, bilaterality
amphitheater, amphitheatre
anticreation, antireaction
anticreative, antireactive
anticrotalic, calcitration
antiparticle, antiprelatic
arsenicating, ascertaining
cacodaemonic, cacodemoniac
         272024758 function calls in 81.687 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
 86778288   38.396    0.000   55.365    0.000 <ipython-input-1-fb9a8026ab34>:2(is_anagram)
        1   26.309   26.309   81.686   81.686 <ipython-input-2-6cc1fa1e5092>:1(main)
        1    0.001    0.001   81.687   81.687 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 _bootlocale.py:11(getpreferredencoding)
        1    0.000    0.000    0.000    0.000 codecs.py:260(__init__)
        1    0.000    0.000    0.001    0.001 cp1252.py:22(decode)
       34    0.000    0.000    0.001    0.000 

<h4>Profiling Result</h4>
<img src="profiling-result.jpg" width="100%">
<br><br>
<h4>Highlights:</h4>
<ul>
    <li>To shorten the processing time, the <b>words_incomplete.txt</b> has been shortened and it consists of only the first 50,000 lines of <b>words_alpha.txt</b><br>&nbsp;<i>$ head words_alpha.txt -n 50000 > words_incomplete.txt </i><br><br></li>    
    <li>As shown in above result, the program completed within 81.686 seconds</li>
    <li>The <b>List.remove()</b> takes 3.196 seconds which is <b>3.9%</b> of the 81.686 seconds</li>
    <li>The <b>len()</b> function takes 13.776 seconds which is <b>19.6%</b> of the 81.686 seconds</li>
    <li>From this profiling we shall optimise <b>len()</b> first before optimizing <b>List.remove()</b></li>
</ul>    